geoips/pydantic/products.py

```diff
@@ -0,0 +1,231 @@
+
+    Args
+    ----
+    plugin_type
+        The type of plugin
+
+    Returns
+    -------
+        list: A list of plugin names for a valid plugin type
+
+    Raises
+    ------
+        AttributeError: If the plugin type is invalid
+
+    """
+    interface_name = plugin_type + "s"
+
+    try:
+        interface = getattr(interfaces, interface_name)
+    except AttributeError:
+        raise AttributeError(f"{plugin_type} is not valid plugin type")
+    return [plugin.name for plugin in interface.get_plugins() or []]
+
+
+def get_plugin_types() -> List[str]:
+    """Retrieve a list of plugin types from available interfaces.
+
+    Returns
+    -------
+    list of str
+        list of possible unique plugin types
+    """
+    interface = []
+    for ifs in interfaces.list_available_interfaces().values():
+        interface.extend(ifs)
+    return list(set(plugin_types[:-1] for plugin_types in interface))
+
+
+class OutputFormatterArgumentsModel(BaseModel):
+    """output_formatters."""
+
+    pass
+
+
+class FilenameFormatterArgumentsModel(BaseModel):
+    """Validate FilenameFormatter arguments."""
+
+    pass
+
+
+class AlgorithmArgumentsModel(BaseModel):
+    """Validate Algorithm arguments."""
+
+    pass
+
+
+class InterpolatorArgumentsModel(BaseModel):
+    """Validate Interpolator arguments."""
+
+    pass
+
+
+class ReaderArgumentsModel(BaseModel):
+    """Validate Reader step arguments."""
+
+    model_config = ConfigDict(extra="forbid")
+    area_def: str = Field(None, description="Area definition identifier.")
+    chans: List[str] = Field(None, description="List of channels to process.")
+    metadata_only: bool = Field(False, description="Flag for metadata-only processing.")
+    self_register: bool = Field(False, description="Flag for self-registration.")
+
+
+class ProductStepDefinitionModel(BaseModel):
+    """Validate step definition : name, arguments."""
+
+    model_config = ConfigDict(extra="forbid")
+    type: str = Field(..., description="plugin type")
+    name: str = Field(..., description="plugin name")
+    arguments: Dict[str, Any] = Field(default_factory=dict, description="step args")
+
+    @model_validator(mode="before")
+    def _validate_plugin_name(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """Validate plugin name."""
+        if not values:
+            raise ValueError("The plugin_name cannot be empty")
+        plugin_name = values.get("name", "").lower()
+        plugin_type = values.get("type", " ").lower()
+
+        valid_plugin_names = get_plugin_names(plugin_type)
+        if plugin_name not in valid_plugin_names:
+            raise ValueError(
+                f"{plugin_name} is invalid. \n\t"
+                f"Must be one of {sorted(valid_plugin_names)}"
+            )
+
+        return values
+
+    @model_validator(mode="before")
+    def _validate_plugin_arguments(cls, values: Dict[str, Any]) -> Dict[str, Any]:
+        """
+        Validate & organize details for each step   .
```

Comment:
spaces?

Reply:
Fixed in 57e51ee890f33ff4111c2bb7bfa702ba78c102fc